src/Certify.Tests/Certify.Core.Tests.Unit/AccessControlTests.cs

```diff
@@ -52,174 +52,806 @@ public Task Add<T>(string itemType, T item)
+            Assert.AreNotEqual(storedSecurityPrinciple.Email, adminSecurityPrinciples[0].Email, $"Expected SecurityPrinciple returned by GetSecurityPrinciple() to not match previous Email '{adminSecurityPrinciples[0].Email}' of SecurityPrinciple passed into AddSecurityPrinciple()");
+            Assert.AreEqual(storedSecurityPrinciple.Email, newSecurityPrinciple.Email, $"Expected SecurityPrinciple returned by GetSecurityPrinciple() to match updated Email '{newSecurityPrinciple.Email}' of SecurityPrinciple passed into AddSecurityPrinciple()");
+        }
+
+        [TestMethod]
+        public async Task TestAccessControlUpdateSecurityPrincipleNoRoles()
+        {
+            var log = new LoggerConfiguration()
+                .WriteTo.Debug()
+                .CreateLogger();
+            var loggy = new Loggy(log);
+            var access = new AccessControl(loggy, new MemoryObjectStore());
+            var contextUserId = "[test]";
+
+            // Add test security principles
+            var adminSecurityPrinciples = new List<SecurityPrinciple> { TestSecurityPrinciples.Admin(), TestSecurityPrinciples.TestAdmin() };
+            adminSecurityPrinciples.ForEach(async p => await access.AddSecurityPrinciple(contextUserId, p, bypassIntegrityCheck: true));
+
+            // Validate email of SecurityPrinciple object returned by AccessControl.GetSecurityPrinciple() before update
+            var storedSecurityPrinciple = await access.GetSecurityPrinciple(contextUserId, adminSecurityPrinciples[0].Id);
+            Assert.AreEqual(storedSecurityPrinciple.Email, adminSecurityPrinciples[0].Email, $"Expected SecurityPrinciple returned by GetSecurityPrinciple() to match Email '{adminSecurityPrinciples[0].Email}' of SecurityPrinciple passed into AddSecurityPrinciple()");
+
+            // Update security principle in AccessControl with a new principle object of the same Id, but different email, with roles undefined
+            var newSecurityPrinciple = TestSecurityPrinciples.Admin();
+            newSecurityPrinciple.Email = "new_test_email@test.com";
+            var securityPrincipleUpdated = await access.UpdateSecurityPrinciple(contextUserId, newSecurityPrinciple);
+            Assert.IsFalse(securityPrincipleUpdated, $"Expected security principle update for {newSecurityPrinciple.Id} to be unsuccessful without roles defined");
+        }
+
+        [TestMethod]
+        public async Task TestAccessControlUpdateSecurityPrincipleBadUpdate()
+        {
+            var log = new LoggerConfiguration()
+                .WriteTo.Debug()
+                .CreateLogger();
+            var loggy = new Loggy(log);
+            var access = new AccessControl(loggy, new MemoryObjectStore());
+            var contextUserId = "[test]";
+
+            // Add test security principles
+            var adminSecurityPrinciples = new List<SecurityPrinciple> { TestSecurityPrinciples.Admin(), TestSecurityPrinciples.TestAdmin() };
```

Comment:
Could we do anything to reduce the number of repeated identical setups per test? I like that the tests are broken out, I just don't like seeing the same setup code over and over. Unless it's varying somehow and I just don't see it.

Reply:
Yes, I think you're right, I had meant to do this sooner! I'll update it to minimize this.

Reply:
Awesome!

Reply:
This is now updated!